### PR TITLE
dcroom cells definition by administrator

### PR DESCRIPTION
**Pull request by orthagh**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #10529

Permits to define cell width and height.
Use case: have a more (or less) compact grid in a dataccenter room view.

### CODE REVIEW

**Review by cconard96 — APPROVED** [review]

File: src/DCRoom.php [hunk 1]
```diff
@@ -158,6 +158,33 @@ public function showForm($ID, array $options = [])
         echo "</td>";
         echo "</tr>";
 
+        echo "<tr class='tab_bg_1'>";
+        echo "<td><label for='dropdown_vis_cell_width$rand'>" . __('Cell width') . "</label></td><td>";
+        Dropdown::showNumber(
+            "vis_cell_width",
+            [
+                'value'  => $this->fields["vis_cell_width"],
+                'min'    => 1,
```
Comment (cedric-anne) [inline src/DCRoom.php:167]: Under 10px, it is not really usable.

```suggestion
                'min'    => 10,
```

File: src/DCRoom.php [hunk 2]
```diff
@@ -158,6 +158,33 @@ public function showForm($ID, array $options = [])
...
+                'rand'   => $rand
+            ]
+        );
+        echo "</td>";
+        echo "<td><label for='dropdown_vis_cell_height$rand'>" . __('cell height') . "</label></td><td>";
+        Dropdown::showNumber(
+            "vis_cell_height",
+            [
+                'value'  => $this->fields["vis_cell_height"],
+                'min'    => 1,
```
Comment (cedric-anne) [inline src/DCRoom.php:179]: Under 10px, it is not really usable.

```suggestion
                'min'    => 10,
```

File: src/DCRoom.php [hunk 3]
```diff
@@ -158,6 +158,33 @@ public function showForm($ID, array $options = [])
         echo "</td>";
         echo "</tr>";
 
+        echo "<tr class='tab_bg_1'>";
+        echo "<td><label for='dropdown_vis_cell_width$rand'>" . __('Cell width') . "</label></td><td>";
+        Dropdown::showNumber(
+            "vis_cell_width",
+            [
+                'value'  => $this->fields["vis_cell_width"],
```
Comment (cedric-anne) [inline src/DCRoom.php:166]: Default value should be proposed on creation.

```suggestion
                'value'  => $this->fields["vis_cell_width"] ?: 40,
```

File: src/DCRoom.php [hunk 4]
```diff
@@ -158,6 +158,33 @@ public function showForm($ID, array $options = [])
...
+                'step'   => 1,
+                'rand'   => $rand
+            ]
+        );
+        echo "</td>";
+        echo "<td><label for='dropdown_vis_cell_height$rand'>" . __('cell height') . "</label></td><td>";
+        Dropdown::showNumber(
+            "vis_cell_height",
+            [
+                'value'  => $this->fields["vis_cell_height"],
```
Comment (cedric-anne) [inline src/DCRoom.php:178]: Default value should be proposed on creation.

```suggestion
                'value'  => $this->fields["vis_cell_height"] ?: 40,
```

File: install/migrations/update_10.0.x_to_10.1.0/dcroom.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    );
+}
+if (!$DB->fieldExists("glpi_dcrooms", "vis_cell_height")) {
+    $migration->addField(
+        "glpi_dcrooms",
+        "vis_cell_height",
+        "int",
+        [
+            'after'   => "vis_cell_width",
+            'value'   => 39,
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/dcroom.php:57]: Maybe a new default to `40` would be less disturbing. Using `update` would permit to keep previous size on existing items.
```suggestion
            'value'   => 40,
            'update'  => 39,
```

**Comment by cedric-anne** [comment]
I noticed 2 bugs.

1) The `+` icon displayed when hovering the gris is not centered on corresponding cell. If cell is small (e.g. 10x10), the `+` icon is displayed on another cell.

2) When using high values, grid may overflow the container.

File: install/migrations/update_10.0.x_to_10.1.0/dcroom.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    );
+}
+if (!$DB->fieldExists("glpi_dcrooms", "vis_cell_height")) {
+    $migration->addField(
+        "glpi_dcrooms",
+        "vis_cell_height",
+        "int",
+        [
+            'after'   => "vis_cell_width",
+            'default' => 40,
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/dcroom.php:63]: Existing entries should use the 39 legacy value.
```suggestion
            'default' => 40,
            'update'  => 39,
```

File: install/migrations/update_10.0.x_to_10.1.0/dcroom.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+if (!$DB->fieldExists("glpi_dcrooms", "vis_cell_width")) {
+    $migration->addField(
+        "glpi_dcrooms",
+        "vis_cell_width",
+        "int",
+        [
+            'after'  => "vis_rows",
+            'update' => 40,
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/dcroom.php:52]: No need to update existing entries, they will use the default value defined by `value`.
```suggestion
```

File: install/migrations/update_10.0.x_to_10.1.0/dcroom.php [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    );
+}
+if (!$DB->fieldExists("glpi_dcrooms", "vis_cell_height")) {
+    $migration->addField(
+        "glpi_dcrooms",
+        "vis_cell_height",
+        "int",
+        [
+            'after'  => "vis_cell_width",
+            'update' => 40,
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/dcroom.php:64]: Existing entries should use the legacy value.
```suggestion
            'update' => 39,
```

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**